### ISSUE TITLE
Image | Tensor union 型を返す関数群の overload を整備

### ISSUE THREAD

**Issue by kurorosu** [kurorosu/pochidetection#619]
## Summary

`apply_letterbox` などが `PIL.Image | torch.Tensor` の union 入力 → union 出力で型推論が崩れている. `@overload` で `Image -> Image` / `Tensor -> Tensor` に分岐し, 呼び出し側 (`core/preprocess.py` / `pipelines/*_pipeline.py` / `utils/infer_debug.py` 等) の代入エラーを解消する.

## Background

`pochidetection/core/letterbox.py::apply_letterbox` は実装が `v2.functional.resize` ベースで PIL / Tensor 多態に対応するが, シグネチャが `Image | Tensor -> Image | Tensor` のため, 呼び出し側で `Tensor` 変数に代入しようとすると mypy が Liskov 違反として弾く. テスト側でも `Image | Tensor` を indexable / `.shape` で扱おうとして union-attr エラー.

## Current Implementation

`uv run mypy .` で本カテゴリは ~10 件:

```
pochidetection\core\letterbox.py:108: error: Argument 1 to "resize" has incompatible type "Image | Tensor"; expected "Tensor"  [arg-type]
pochidetection\core\preprocess.py:50: error: Incompatible types in assignment (expression has type "Image | Tensor", variable has type "Tensor")  [assignment]
pochidetection\pipelines\rtdetr_pipeline.py:165: error: Incompatible types in assignment (expression has type "Image | Tensor", variable has type "Image | ndarray[...]")  [assignment]
pochidetection\pipelines\ssd_pipeline.py:157: error: Incompatible types in assignment (expression has type "Image | Tensor", variable has type "Image | ndarray[...]")  [assignment]
pochidetection\utils\infer_debug.py:109: error: Item "Tensor" of "Image | Tensor" has no attribute "save"  [union-attr]
tests\test_core\test_letterbox.py:124-150: error: Value of type "Image | Tensor" is not indexable / has no attribute "shape"
```

## Proposed Changes

`apply_letterbox` を `@overload` で型分岐させる.

```python
# pochidetection/core/letterbox.py
from typing import overload
from PIL.Image import Image
import torch

@overload
def apply_letterbox(image: Image, params: LetterboxParams, pad_value: int = 0) -> Image: ...
@overload
def apply_letterbox(image: torch.Tensor, params: LetterboxParams, pad_value: int = 0) -> torch.Tensor: ...
def apply_letterbox(image, params, pad_value=0):
    ...
```

呼び出し側は変更不要 (overload で型推論される). ただし `core/preprocess.py:50` 周辺は `tensor_uint8: Tensor` を渡しているので overload 解決で `Tensor` が返るようになる. テスト側も同様.

`utils/infer_debug.py:109` は引数が PIL 限定の場面なので, 関数シグネチャを `Image` 限定に絞るか, isinstance 分岐を入れる. 個別判断.

## Acceptance Criteria

- `uv run mypy .` で `core/letterbox.py` / `core/preprocess.py` / `pipelines/rtdetr_pipeline.py:165` / `pipelines/ssd_pipeline.py:157` / `utils/infer_debug.py:109` / `tests/test_core/test_letterbox.py` の本カテゴリエラーが 0 件
- 既存テスト (`uv run pytest`) が pass する
- 動作変更なし (overload は型注釈のみの変更)

## Labels

`refactor`

## Branch

`refactor/`
